### ISSUE TITLE
Remove Thin Client from storage stage

### ISSUE THREAD

**Issue by t-neIson** [solana-labs-community/solana#47]
#### Problem
Storage stage uses a thin client to talk to the fullnode it's running within. This doesn't seem that efficient and results in many errors on nodes that haven't setup local RPC services.

#### Summary of Changes
Removed thin client from storage stage entirely. Instead use a bank forks and a udpsocket to send txs to the leader via the node's own tpu. This also reduced occurrences of storage accounts not showing up on validators some times.

**Comment by sakeridge** [comment]
and I talked about the storage mining part being separable from the validation part, this kind of ties them together since you need to be a fullnode to have the bank_forks. What is more efficient about not using thin_client? This seems more inefficient with maybe slightly more forwarding but likely a wash overall. Better filtering for rpc nodes could reduce the errors. Anyway, I'm not opposed to the change, but somethings to think about.

How does this deal with failures? I would think there is not a retry on the tpu forwarder, is that right?

**Comment by t-neIson** [comment]
Unsure how this is dealing with failures. Something about rpc seems flaky right now, looking into that separately(could boil down to all this working bank stuff on bank forks). Somehow this is more reliably able to setup accounts and submit proofs advertise blockhashes than via rpc. I'm not seeing odd AccountNotFound errors as frequently.

Anyway, filtering based on local rpc controls storage stage implicitly, we ought to have some explicit switch for storage verification some day. I agree that this ties storage validation into fullnode even more but while it lives inside a fullnode it really shouldn't be rpc'ing back to itself, it's just unnecessary it has everything it needs to work without that. Also it doesn't add any more forwarding than the rpc mechanism since rpc will end up using forwarding anyway.

How is this tying mining and validation together? it's just tying validation into the fullnode it runs in.

**Comment by sakeridge** [comment]
```rust
            .spawn(move || {
                let transactions_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
                loop {
                    match tx_receiver.recv_timeout(Duration::from_secs(1)) {
```
can you file an issue for "extra stage in storage thingie", and just send the TXes to the TPU directly instead?

**Comment by mvilnes** [comment]
#3986
It was doing send/confirm/retry. So didn't want to wait around doing that in the main storage loop.

**Comment by CrieasofCarrots** [comment]
fine, but more work downstream, yeah?

**Comment by jeffweshington** [comment]
Definitely.

**Comment by t-neIson** [comment]
Right, so you might have a separate storage mining machine with a bunch of different hardware specialized for that which is different than your validator mining machine. The storage mining machine would not want to be a full node, it would want to just ask the validator node the minimum set of questions to get the data it needs to do the storage mining validation.

**Comment by sakeridge** [comment]
Hmm I don't think I get it. I mean I understand that a storage mining machine is separate from a Storage Validation machine (aka just a validator?). But This doesn't impact the storage mining machine at all, it's purely contained in the validator.

Yup, makes sense. This PR doesn't change that at all no?

What am I missing :S

**Comment by mvilnes** [comment]
I mean the transaction validator machine could be different than the storage mining validator machine (not the replicator).

**Comment by CrieasofCarrots** [comment]
Ah okay. That makes much more sense. Yeah I didn't really make it that much worse. When and if we pull it out, we'll just run a replay stage without voting and it's good to go.
